knowledge/pkg/datastore/documentloader/ocr/openai/openai.go

```diff
@@ -223,3 +242,74 @@ func (o *OpenAIOCR) sendImageToOpenAI(base64Image string) (string, error) {
 
 	return result.Choices[0].Message.Content, nil
 }
+
+func requestWithExponentialBackoff(ctx context.Context, client *http.Client, req *http.Request, maxRetries int, handleRateLimit bool) ([]byte, error) {
+
+	const baseDelay = time.Millisecond * 200
+	var resp *http.Response
+	var err error
+
+	var failures []string
+
+	// Save the original request body
+	var bodyBytes []byte
+	if req.Body != nil {
+		bodyBytes, err = io.ReadAll(req.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read request body: %v", err)
+		}
+	}
+
+	for i := 0; i < maxRetries; i++ {
+		// Reset body to the original request body
+		if bodyBytes != nil {
+			req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+		}
+
+		resp, err = client.Do(req)
+		if err == nil && resp.StatusCode == http.StatusOK {
+			if resp.Body == nil {
```

Comment:
nit: not possible for `resp.Body` to be `nil` if `err == nil`.

Reply:
TIL :)